modules/aggregation-layers/src/heatmap-layer/triangle-layer.ts

```diff
@@ -33,60 +33,86 @@ type _TriangleLayerProps = {
   vertexCount: number;
   colorTexture: Texture;
   maxTexture: Texture;
-  texture: Texture;
+  weightsTexture: Texture;
 };
 
 export default class TriangleLayer extends Layer<_TriangleLayerProps> {
   static layerName = 'TriangleLayer';
 
   state!: {
     model: Model;
+    positions: Buffer,
+    texCoords: Buffer,
   };
 
   getShaders() {
-    return {vs, fs, modules: [project32]};
+    // TODO(donmccurdy): Including 'picking' fixes errors, but I'm not sure
+    // heatmap-layer really supports picking?
+    return {vs, fs, modules: [project32, picking]};
   }
 
   initializeState({device}: LayerContext): void {
+    console.log('TriangleLayer#initializeState()'); // TODO(donmccurdy): DO NOT SUBMIT.
+    this.setState({model: this._getModel(device)});
+  }
+
+  _getModel(device: Device): Model {
+    console.log('TriangleLayer#_getModel()'); // TODO(donmccurdy): DO NOT SUBMIT.
+
     const attributeManager = this.getAttributeManager()!;
+    const {vertexCount, data, weightsTexture, maxTexture, colorTexture} = this.props;
+
+    // TODO(donmccurdy): This is probably doing nothing unless passed to the Model?
     attributeManager.add({
       positions: {size: 3, noAlloc: true},
       texCoords: {size: 2, noAlloc: true}
     });
-    this.setState({
-      model: this._getModel(device)
-    });
-  }
 
-  _getModel(device: Device): Model {
-    const {vertexCount} = this.props;
+    debugFBO(weightsTexture, {id: 'triangle-weightsTexture', opaque: true, top: '0', rgbaScale: 255});
+
+    // TODO(donmccurdy): cleanup.
+    console.log({
+      // weightsTexture: weightsTexture.
+      triangleAttributes: (Object.entries((data as any).attributes) as any)
+        .map(([name, buffer]: [string, Buffer]) => [name, getBufferData(buffer, Float32Array)])
+    });
 
     return new Model(device, {
-      ...this.getShaders(),
+      ...(this.getShaders() as any),
       id: this.props.id,
-      geometry: new Geometry({
-        topology: 'triangle-fan-webgl',
-        vertexCount
-      })
+      bindings: {weightsTexture, maxTexture, colorTexture}, // TODO(donmccurdy): required?
+      attributes: (data as any).attributes, // TODO(donmccurdy): types.
+      bufferLayout: [
+        {name: 'positions', format: 'float32x3'},
+        {name: 'texCoords', format: 'float32x2'},
+      ],
```

Comment:
TODO:

- [ ] use `this.getAttributeManager().getBufferLayouts()`
- [ ] avoid `triangle-fan-webgl`